pkg/vault/contracts/CompositeLiquidityRouter.sol

```diff
@@ -0,0 +1,512 @@
+                )
+            ),
+            (address[], uint256[])
+        );
+    }
+
+    function removeLiquidityProportionalFromNestedPoolsHook(
+        RemoveLiquidityHookParams calldata params
+    ) external nonReentrant onlyVault returns (address[] memory tokensOut, uint256[] memory amountsOut) {
+        IERC20[] memory parentPoolTokens = _vault.getPoolTokens(params.pool);
+
+        (, uint256[] memory parentPoolAmountsOut, ) = _vault.removeLiquidity(
+            RemoveLiquidityParams({
+                pool: params.pool,
+                from: params.sender,
+                maxBptAmountIn: params.maxBptAmountIn,
+                minAmountsOut: new uint256[](parentPoolTokens.length),
+                kind: params.kind,
+                userData: params.userData
+            })
+        );
+
+        for (uint256 i = 0; i < parentPoolTokens.length; i++) {
+            address childPool = address(parentPoolTokens[i]);
+
+            if (_vault.isPoolRegistered(childPool)) {
+                // Token is a BPT, so remove liquidity from the child pool.
+
+                // We don't expect the sender to have BPT to burn. So, we flashloan tokens here (which should in
+                // practice just use existing credit).
+                _vault.sendTo(IERC20(childPool), address(this), parentPoolAmountsOut[i]);
+
+                IERC20[] memory childPoolTokens = _vault.getPoolTokens(childPool);
+                // Router is an intermediary in this case. The Vault will burn tokens from the router, so Router is
+                // both owner and spender (which doesn't need approval).
+                (, uint256[] memory childPoolAmountsOut, ) = _vault.removeLiquidity(
+                    RemoveLiquidityParams({
+                        pool: childPool,
+                        from: address(this),
+                        maxBptAmountIn: parentPoolAmountsOut[i],
+                        minAmountsOut: new uint256[](childPoolTokens.length),
+                        kind: params.kind,
+                        userData: params.userData
+                    })
+                );
+                // Return amounts to user.
+                for (uint256 j = 0; j < childPoolTokens.length; j++) {
+                    _currentSwapTokensOut().add(address(childPoolTokens[j]));
+                    _currentSwapTokenOutAmounts().tAdd(address(childPoolTokens[j]), childPoolAmountsOut[j]);
+                }
+            } else {
+                // Token is not a BPT, so return the amount to the user.
+                _currentSwapTokensOut().add(childPool);
+                _currentSwapTokenOutAmounts().tAdd(childPool, parentPoolAmountsOut[i]);
+            }
+        }
+
+        // The hook writes current swap token and token amounts out.
+        // We copy that information to memory to return it before it is deleted during settlement.
+        tokensOut = InputHelpers.sortTokens(_currentSwapTokensOut().values().asIERC20()).asAddress();
```

Comment:
Maybe we don't need to sort the tokens out. Just returning the array of token addresses and output amounts should do.

Reply:
Done! As discussed, the user will pass tokensOut array, since he needs to know the token addresses beforehand to pass the minAmountsOut correctly.